nodeJS/express/views.md

```diff
@@ -0,0 +1,180 @@
+
+Here's a quick example that makes includes arrays and loop logic.
+
+```javascript
+<% const animals = ["Cat", "Dog", "Lemur", "Hawk"] %>
+
+<% animals.map(animal => { %>
+- <%= animal%>s are cute <% }) %>
+```
+
+### Using EJS with Express
+
+Let's use EJS with Express. First, create an EJS template file called `index.ejs` in the `views` subdirectory, and add the following:
+
+```html
+<html>
+  <body>
+    <%= message %>
+  </body>
+</html>
+```
+
+And in app.js, render this template file in one of your routes:
+
+```javascript
+app.get("/", (req, res) => {
+  res.render("index", { message: "EJS rocks!" });
+});
+```
+
+Start the server and go to the `/` route in the browser. You should see:
+
+```html
+EJS rocks!
+```
+
+If you inspect the HTML in the browser's dev tools, you can see HTML is structures exactly like how we wrote the EJS template with the `message` variable replaced with its value.
+
+When you hit the `/` route, `res.render("index", { message: "EJS rocks!" });`  is the line that sends back the response. Since we've already defined the `views` and `view engine` app properties, the first argument of `res.render` is programmed to look for "a template called index in the specified folder", while the second argument is an object of variables that are to be made available to that specific template.
+
+### Reusable templates
+
+You may want to include webpage components that are shared across different pages, such as a sidebar or a header. To insert such components into your pages, we make use of the `include` command. This requires the name of the file to be inserted, and optionally an object of data you wish to pass.
+
+Say you have the following navbar component called `"navbar.ejs"`:
+
+```html
+<!-- navbar.ejs -->
+<nav>
+  <ul>
+    <% for (let i = 0; i < links.length; i++) { %>
+      <li>
+        <a href="<%= links[i].href %>">
+          <span> <%= links[i].text %> </span>
+        </a>
+      </li>
+    <% } %>
+  </ul>
+</nav>
+```
+
+You can insert this component into another EJS file like so:
+
+```html
+<!-- index.ejs -->
+<html>
+  <head>
+    <title>Homepage</title>
+    <body>
+      <%- include('navbar', {links: links}) %>
+    </body>
+  </head>
+</html>
+```
+
+This can be used to include headers and footers in all of your pages, for example.
+
+Note that the navbar expects a `links` value. To pass this data into the navbar, you can pass it when rendering `index.ejs` which contains the navbar. Modify `app.js` such that a `links` object is defined and passed into the `render` function in the `"/"` route handler:
+
+```javascript
+// app.js
+const links = [
+  { href: "/", text: "Home" },
+  { href: "about", text: "About" },
+];
+
+app.get("/", (req, res) => {
+  res.render("index", { links: links });
+});
+```
+
+Here's another example of how to use `includes` to dynamically render a list of variables:
+
+```html
+<ul>
+  <% users.map( user => { %> <%- include('user/show', {user: user}) %> <% }) %>
+</ul>
```

Comment:
```suggestion
<ul>
  <% users.forEach((user) => { %>
    <%- include('user/show', {user: user}); %>
  <% }); %>
</ul>
```
This example would be better written as per the [EJS docs `includes` example](https://ejs.co/#:~:text=whitespace%20after%20it-,Includes,-Includes%20are%20relative), just with an arrow function instead.
The split lines is easier to read than multiple `<% %>` in the same line, and switching to `forEach` instead of `map` as we're not using the return value for anything - just including as we iterate over the array.

Looking into how we can get `ejs` syntax highlighting like we do for `erb`, as the markdown preview tool seems to suggest `ejs` is not (yet) supported. If and when we can, then we can use `ejs` instead of `html` for better colours.

Reply:
Thanks for the suggestions Mao! I'll implement your suggestion, and change any blocks that require the `ejs` tag.